### PR TITLE
2.0.2

### LINKED ISSUE #210: Infer SecurityDomain from filter [Security Middleware]

### LINKED ISSUE #235: Add support for custom directives.
### Discussed in https://github.com/twinlogix/typetta/discussions/232

<div type='discussions-op-text'>

<sup>Originally posted by **Atlinx** June  1, 2022</sup>
Although we already have the `@schema(metadata: [])` decorator, it would nice if we could write our own decorators to make our gql schema less verbose. These decorators could be placed under the existing metadata field, or in it's own unique field.

**Ex.** 
GraphQL
```graphql
type User @entity @mongodb {
  id: ID! @id(from: "db") @alias(value: "_id")
  createdAt: Date! @createdAt # This is a custom decorator
  name: Date! @deprecated(reason: "Use firstName + lastName") # This is a custom decorator
  firstName: String
  lastName: String
  @@unique(fields: ["firstName", "lastName"]) # This is a custom decorator
}
```
Schema
```typescript
export function userSocialSchema(): T.Schema<types.Scalars> {
  return {
    id: {
      type: 'scalar',
      scalar: 'ID',
      isId: true,
      generationStrategy: 'db',
      required: true,
      alias: '_id',
    },
    createdAt: {
      type: 'scalar',
      scalar: 'String',
      required: true,
      directives: [
        {
          name: "createdAt",
        }
      ]
    },
    name: {
      type: 'scalar',
      scalar: 'String',
      required: true,
      directives: [
        {
          name: "deprecated",
          args: {
            reason: "Use firstName + lastName"
          }
        }
      ]
    },
    firstName: {
      type: 'scalar',
      scalar: 'String',
    },
    lastName: {
      type: 'scalar',
      scalar: 'String',
    },
    __directives: {
      {
        name: "unique",
        args: {
          fields: [ "firstName", "lastName" ]
        }
      }
    }
  }
}
```</div>

### LINKED ISSUE #258: Document `resolveRelations`

### LINKED ISSUE #261: Update operation security domain documentation
When #210 is completed the operation security domain can be inferred from filter or specified as array and not anymore as a plain security domain.

### PR DESCRIPTION
**Pull request by edobrb**
- [x] improve create resolver
- [x] infer operation security domain (closes #261 closes #210)
- [x] document resolveRelations (closes #258)
- [x] improve middleware typing 
- [x] add the possibility to avoid awaiting log calls 
- [x] custom directives (closes #235)

### CODE REVIEW

**Comment by minox86** [comment]
I think we need some more tests to increase the coverage, at least to 90%.

File: docs/chapters/client/read-write.md [hunk 1]
```diff
@@ -149,6 +150,34 @@ The return of this query is the following object:
 })
 ```
 
+## Resolve Relations
+
+The [resolveRelations](/typedoc/classes/AbstractDAO.html#resolveRelations){:target="_blank"} API allows you resolves some of the relations of a given input. It accepts the return of an insert operation, and resolves the relations based on the projection. It's useful when, after an insert, we want to return the entity within a GraphQL resolver.
```
Comment (minox86) [inline docs/chapters/client/read-write.md:155]:
```suggestion
The [resolveRelations](/typedoc/classes/AbstractDAO.html#resolveRelations){:target="_blank"} API allows you to resolve some of the relations of a given input. It accepts partial entity and resolves the relations based on the given projection. It's mainly useful when, after an insert, you want to return an entity within a GraphQL resolver.
```

File: docs/chapters/client/read-write.md [hunk 2]
```diff
@@ -149,6 +150,34 @@ The return of this query is the following object:
...
+
+Below is an example of a relations resolution:
```
Comment (minox86) [inline docs/chapters/client/read-write.md:157]: Following an example of relations resolution:

**Review by minox86 — CHANGES_REQUESTED** [review]

**Comment by edobrb** [comment]
image

**Comment by minox86** [comment]
image

**Pull request merged**